### PR TITLE
aes-gcm-avx512.pl: fix non-reproducibility issue

### LINKED ISSUE #20954: OpenSSL is non-reproducible
I am building dependencies from FDSK.

I am using this [git_tag](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/tree/bb8337768b0a492431883cf73e09986eac976719) of FDSK for my project, which happens to reference this [version](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/blob/bb8337768b0a492431883cf73e09986eac976719/elements/components/openssl.bst#L69) of openssl.

My issue is that openssl reports itself to be non-reproducible. From my understanding, it is because of this [file](https://github.com/openssl/openssl/blob/a92271e03a8d0dee507b6f1e7f49512568b2c7ad/crypto/modes/asm/aes-gcm-avx512.pl#L195), which happens to introduce random strings.

I need to know if I add a patch file to my system to seed the random number generator, if this would have potential security flaws?

For further info, please read: https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/issues/1589

### PR DESCRIPTION
**Pull request by trigpolynom**
Fix #20954. Made changes suggested in [this](https://github.com/openssl/openssl/issues/20954#issuecomment-1547682819) comment. Passed all tests.

CLA: trivial

### CODE REVIEW

File: crypto/modes/asm/aes-gcm-avx512.pl [hunk 1]
```diff
@@ -417,7 +411,7 @@ sub PROLOG {
 sub EPILOG {
   my ($hkeys_storage_on_stack, $payload_len) = @_;
 
-  my $rndsuffix = &random_string();
+  my $rndsuffix = "aesgcmavx512_" . $count++;
```
Comment (levitte) [inline crypto/modes/asm/aes-gcm-avx512.pl:414]: Considering that it's just a suffix on local labels, this should be sufficient:

```suggestion
  my $rndsuffix = $count++;
```

Comment (levitte) [reply crypto/modes/asm/aes-gcm-avx512.pl:414]: That is, of course, valid for all the spots where you made this change, not just here.

File: crypto/modes/asm/aes-gcm-avx512.pl [hunk 2]
```diff
@@ -200,14 +200,8 @@
...
 
-# ; Generates "random" local labels
-sub random_string() {
-  my @chars  = ('a' .. 'z', 'A' .. 'Z', '0' .. '9', '_');
-  my $length = 15;
-  my $str;
-  map { $str .= $chars[rand(33)] } 1 .. $length;
-  return $str;
-}
+# ; Counter variable that replaces previous usage of random_string()
```
Comment (mspncp) [inline crypto/modes/asm/aes-gcm-avx512.pl:203]: Historical comments belong in the commit message, not the code comment ;-). This information is meaningless for the current implementation.
```suggestion
```

**Review by mspncp — CHANGES_REQUESTED** [review]
Hello @trigpolynom,

thank you for your contribution. I just have some cosmetical change requests:

Your commit message

```
CLA: trivial
```

and the title of your pull request

> Fixes #20954 (non-reproducibility issue)

are suboptimal in several respects:

- The subject line of the commit message should give a meaningful summary of the change
- The `CLA: trivial` and the `Fixes #xyz` lines belong into the *body* of the commit message, not the *subject line* (the body is separated from the subject line by an empty line)
- If the pull request consists of a single commit, GitHub suggests the subject line as pull request title and the commit body as pull request description. This is a good choice, I'd recommend to follow GitHub's advice  (You may add additional information of course.). 


The following commit message is just a suggestion to give you an idea, feel free to change the wording to your liking.

```
aes-gcm-avx512.pl: fix non-reproducibility issue

Replace the random suffix with a counter, to make the
build reproducible.

Fixes #20954

CLA: trivial
```

And finally, please change your pull request title to match your subject line:

> aes-gcm-avx512.pl: fix non-reproducibility issue

**Comment by mspncp** [comment]
**Note** the fact that the `CLA: trivial` annotation is in the subject line is the reason why the bot is  still complaining about a missing CLA. 

A propos CLA:  generally, we set a very low bar to when a pull request becomes non-trivial. (To avoid repeating the same legal discussions over and over, since we are no lawyers.) Only trivial one-line-fixes or typographical corrections are considered trivial. Given the  fact that this is not your first contribution and that the change is borderline, I'd prefer if you could sign and send us a [Contributor License Agreement (CLA)](https://www.openssl.org/policies/cla.html). Also as a motivation for future, non-trivial contributions from you. :-)

File: crypto/modes/asm/aes-gcm-avx512.pl [hunk 3]
```diff
@@ -200,14 +200,8 @@
...
-# ; Generates "random" local labels
-sub random_string() {
-  my @chars  = ('a' .. 'z', 'A' .. 'Z', '0' .. '9', '_');
-  my $length = 15;
-  my $str;
-  map { $str .= $chars[rand(33)] } 1 .. $length;
-  return $str;
-}
+# ; Counter variable that replaces previous usage of random_string() 
+my $count = 0;
```
Comment (tom-cosgrove-arm) [inline crypto/modes/asm/aes-gcm-avx512.pl:204]: Please don't call a variable `$count` except in trivial scripts. How about `$label_count`?

File: crypto/modes/asm/aes-gcm-avx512.pl (hunk 1 above)
Comment (tom-cosgrove-arm) [reply crypto/modes/asm/aes-gcm-avx512.pl:414]: Can I go further, and ask that this no longer be called `$rndsuffix` but either just `$suffix` or `$label_suffix`?

Let's try to have the code make as much sense in its new form as possible

**Comment by t8m** [comment]
This is definitely not something acceptable with CLA: trivial. So I am glad you're willing to sign a regular CLA.

File: crypto/modes/asm/aes-gcm-avx512.pl (hunk 2 above)
Comment (trigpolynom) [reply crypto/modes/asm/aes-gcm-avx512.pl]: deleted comment

**Comment by trigpolynom** [comment]
submitted CLA

File: crypto/modes/asm/aes-gcm-avx512.pl (hunk 3 above)
Comment (trigpolynom) [reply crypto/modes/asm/aes-gcm-avx512.pl]: made change

**Comment by trigpolynom** [comment]
Something is causing a build error. looking into it

**Review by levitte — APPROVED** [review]
Subject to CLA resolution, of course

File: crypto/modes/asm/aes-gcm-avx512.pl [hunk 4]
```diff
@@ -200,14 +200,7 @@
...
 
-# ; Generates "random" local labels
-sub random_string() {
-  my @chars  = ('a' .. 'z', 'A' .. 'Z', '0' .. '9', '_');
-  my $length = 15;
-  my $str;
-  map { $str .= $chars[rand(33)] } 1 .. $length;
-  return $str;
-}
+my $label_count = 0;
```
Comment (tom-cosgrove-arm) [inline crypto/modes/asm/aes-gcm-avx512.pl:203]: (minor; not sure if it's worth another round to fix this; up to you) variable declaration and initialisation probably shouldn't be under "Helper functions"

**Comment by tom-cosgrove-arm** [comment]
I will approve once CLA is fixed (which I think will need rebase to remove `CLA: trivial` from first commit message)

**Comment by mspncp** [comment]
@trigpolynom your pull request now consists of three commits. Those need to be squashed into a single commit, keeping only the commit message of the [second commit](https://github.com/openssl/openssl/pull/22415/commits/5acbed119a6e345b1eb930c2fc642f4e02b0eeb9) (with the `CLA: non-trivial` removed).

Typically, this squashing is done with an interactive rebase (`git rebase -i --keep-base master`), followed by a force-push (`git push --force-with[-lease]`). Do you know how to do it yourself? If not, you have two options:

1. ask me to do the squashing for you
2. ask me to explain the steps to you and try it yourself

I you choose step 2, don't worry about messing it up: everything can be undone in Git and I will help you if you get in troubles ;-)

**Comment by trigpolynom** [comment]
@mspncp I'll fix it tonight myself. I swear this isn't my first rodeo, despite how much it seems to be. Just anxious to get my first contribution in here.

**Comment by mspncp** [comment]
Sure, we're not in a hurry. And I know from my own experience that it is much more satisfying for you if you are able to fix the problems by yourself.

**Comment by trigpolynom** [comment]
@mspncp Okay i think ive finally got the commits in check

File: crypto/modes/asm/aes-gcm-avx512.pl (hunk 4 above)
Comment (trigpolynom) [reply crypto/modes/asm/aes-gcm-avx512.pl:203]: just saw this. making change now

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM - thanks

**Review by mspncp — APPROVED** [review]

**Comment by trigpolynom** [comment]
hi everyone. when will this potentially be merged?

**Comment by mspncp** [comment]
The repository is currently frozen in preparation of the upcoming 3.2 beta release. But it will certainly be merged shortly after that and before the next service release.

**Comment by trigpolynom** [comment]
@mspncp okay sounds good. just excited. Thanks!

**Comment by hlandau** [comment]
Adding 3.2 on the basis that any approved PR targeting master and at least one other branch should be assumed a candidate for 3.2.

**Comment by hlandau** [comment]
Merged to master, 3.1 and 3.2. Thank you.

This did not merge cleanly to 3.0; you may wish to open another PR for 3.0.

**Pull request closed without merging**